### DISCUSSION TITLE
How to schedule simple commands?

### DISCUSSION THREAD

**Discussion by akauppi** [zwave-js/zwave-js-ui#3509] [Q&A] [answered]
I am slightly embarrassed to be asking this.. It's perfectly fine that the project doesn't aim to schedule device actions, but.. here goes.

I've got ZWave-JS-UI running and my devices are recognized. I'd like to schedule simple "power on/off" actions for some smart plugs. Is there any way I could do this without resorting to whole Home Assistant? Having databases and what not it looks way heavier than what I would need for: keep lights on between 10am .. 10pm.

**Comment by robertsLando** [comment]
Yes, you can: https://zwave-js.github.io/zwave-js-ui/#/usage/setup?id=scheduled-jobs

Basically you can schedule custom [driver functions](https://zwave-js.github.io/zwave-js-ui/#/usage/driver_function?id=driver-function) to run,

**Comment by akauppi** [comment]
Thanks.

That means technically it could be done. But the developer experience wasn't great (I don't see a cursor, and get no logs out though I tried). It would be coding in dark, and without API docs, I guess. Cannot switch the bulb on.

What I do see is a "targetId" (2-37-0, see below). If you have a snippet that switches such on/off, I can give it a try.

![image](https://github.com/zwave-js/zwave-js-ui/assets/1297734/525d77f2-e2d8-46e2-bd20-02192e42b312)

On my own, I'll go and make some Rust automation to do this.

**Comment by robertsLando** [comment, marked as answer]
Here is your driver function to write `true` on that specific switch:

```js
const zwaveClient = this.zwaveClient

const vId = {
    nodeId: 2,
    commandClass: 37,
    endpoint: 0,
    property: 'targetValue',
}

zwaveClient.writeValue(vId, true)
```

**Reply by akauppi** [reply]
This is more than I hoped for.  

..but it did not work.

![image](https://github.com/zwave-js/zwave-js-ui/assets/1297734/c8c1f48c-acd7-4c61-977f-84892fc7925a)

I thought running it every minute would be a good way to debug. The switch hasn't turned on a single time - though manually I can flip it.

I cannot really work with the scheduling UX anyways:

- I am not able to see output/any logs when this code runs (or even confirm it does). Looked in the Debug tab.
- I am not able to edit the scheduled entries. This may be a UI bug - when I go edit, there's no "SAVE" button - only "EDIT".  Thus my workflow becomes a loop of make-a-new-entry; remove; make-new...

**Reply by robertsLando** [reply]
You can add some logs:

```js
const { logger, zwaveClient } = this

const vId = {
    nodeId: 2,
    commandClass: 37,
    endpoint: 0,
    property: 'targetValue',
}

logger.info("Running scheduled write...")
zwaveClient.writeValue(vId, true)
```

Anyway you should see on logs the write request like `Writing xxx to yyy`, did you tried to keep the debug window opened?

**Reply by akauppi** [reply]
This is what I tried for logging. But it doesn’t work for me. There’s likely some local glitch in my end and I don’t intend to go deeper after it. In order to debug, we would need to screen-share and what not. Not worth it, if it works for you.

Thanks anyhow.  Will keep using zwave-js-ui, just not for scheduling!

**Reply by robertsLando** [reply]
Did you enabled log to files on settings? Is log level set to debug?

**Reply by akauppi** [reply]
That's one more mystery right there. These are my General settings:

![image](https://github.com/zwave-js/zwave-js-ui/assets/1297734/172a77b7-1a6b-41d3-b7f9-4af38cfe0cf4)

I understand "log to file" logs to the file system. I want to see logs in the debug, so it's off.

However, on the Debug page:

![image](https://github.com/zwave-js/zwave-js-ui/assets/1297734/a164abc2-8742-44bd-864c-137dc1dbf63b)

The logs I see here are from a Docker health check that pings the main page.

I don't get what's going on here, so will continue looking at Rust as an easier ;P option. Yeah, right.

**Reply by robertsLando** [reply]
Yeah my question was because if you enable log to file you will not be able to see logs on debug window, also remember there are 2 types of logs, zwave-js logs (the driver that communicate with the stick) and zwave-js-ui logs (the zui application), in order to see both logs there you need to disable log to files for boths, anyway mind to enable them when finished as it is helpful to have logs files when bugs happen, them must be uploaded along with the issues on gh :)

**Reply by akauppi** [reply]
Ok - solved it.

The MQTT Gateway needs to be enabled - otherwise scheduled jobs don't run.

>![image](https://github.com/zwave-js/zwave-js-ui/assets/1297734/dbba72ec-8d4f-42df-940d-e8567286e5d8)

I now get the logs. :) And I can schedule switches on/off successfully.  Mystery fog faded.

**Reply by robertsLando** [reply]
@akauppi Wait, this reminds me an old bug I solved, what version are you using?

**Reply by robertsLando** [reply]
Found it: https://github.com/zwave-js/zwave-js-ui/pull/3409

Has been released more then 2 months ago, I think you should update your instance :)

**Reply by akauppi** [reply]
True. I was using `:latest` Docker image, but it was 3 months old.

From my side, this can now be closed (as an issue), but since it's a discussion I don't know what happens next. 

Thanks for the help - all good now!
